### PR TITLE
Add script to deploy eclipse che in disconnected environment

### PR DESCRIPTION
**Pull request by flacatus**
Signed-off-by: Flavius Lacatusu <flacatus@redhat.com>

### What does this PR do?
This PR add necessary scripts to deploy Eclipse Che in Openshift Disconnected environment.

Ref issue:https://github.com/eclipse/che/issues/19534

### Screenshot/screencast of this PR

N/A
### What issues does this PR fix or reference?

N/A
### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [Nightly OLM bundle is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-nightly-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by flacatus** [comment]
DO NOT REVIEW YET PLEASE... I'm still testing on jenkins

File: .github/bin/resources/slack-message-template.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{   
+    "channel":"che-qe-notifications",
```
Comment (flacatus) [inline .github/bin/resources/slack-message-template.json:2]: I will change to deploy team channel before merge

File: .ci/oci-disconnected.sh [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+cd che-machine-exec
+podman build -f build/dockerfiles/Dockerfile -t "${INTERNAL_REGISTRY_URL}"/"${ORGANIZATION}"/che-machine-exec:"${TAG_NIGHTLY}" .
+cd .. && rm -rf che-machine-exec
+
+# Push che-machine-exec, devfile and plugins image to private registry
+podman push --authfile="${REG_CREDS}" --tls-verify=false "${INTERNAL_REGISTRY_URL}"/"${ORGANIZATION}"/che-devfile-registry:"${TAG_NIGHTLY}"
+podman push --authfile="${REG_CREDS}" --tls-verify=false "${INTERNAL_REGISTRY_URL}"/"${ORGANIZATION}"/che-plugin-registry:"${TAG_NIGHTLY}"
+podman push --authfile="${REG_CREDS}" --tls-verify=false "${INTERNAL_REGISTRY_URL}"/"${ORGANIZATION}"/che-machine-exec:"${TAG_NIGHTLY}"
+
+# Get all containers images used in eclipse-che deployment(postgresql, che-server-che-broket...)
```
Comment (tolusha) [inline .ci/oci-disconnected.sh:122]: Typo?

File: .ci/oci-disconnected.sh [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+export SKIP_TEST=true && ./build.sh --organization "${ORGANIZATION}" \
+           --registry "${INTERNAL_REGISTRY_URL}" \
+           --tag "${TAG_NIGHTLY}" \
+           --offline \
+           --skip-digest-generation
+
+cd .. && rm -rf che-plugin-registry
+
+# Build Che-Machine-Exec
```
Comment (tolusha) [inline .ci/oci-disconnected.sh:111]: Do we need to build machine exec image if  can use nightly one.
https://quay.io/repository/eclipse/che-machine-exec?tag=latest&tab=tags

File: .ci/oci-disconnected.sh [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+# Obtain SIDECAR_IMAGE
+podman run --authfile=${XDG_RUNTIME_DIR}/containers/auth.json -it --rm \
+  --entrypoint cat "${INTERNAL_REGISTRY_URL}"/eclipse/che-plugin-registry:nightly  /var/www/html/v3/plugins/golang/go/latest/meta.yaml > /tmp/workspace.yaml
+
+export SIDECAR_IMAGE=$(cat /tmp/workspace.yaml | yq '.spec.containers[] | .image' -r)
+
+if [[ "$SIDECAR_IMAGE" =~ ^quay.io* ]]; then
+    REGISTRY_IMG_NAME=$(echo $SIDECAR_IMAGE | sed -e "s/quay.io/"${INTERNAL_REGISTRY_URL}"/g")
+    echo $SIDECAR_IMAGE
```
Comment (tolusha) [inline .ci/oci-disconnected.sh:175]: Is it for debug purpose ?

File: .ci/oci-disconnected.sh [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+spec:
+  auth:
+    updateAdminPassword: false
+  server:
+    airGapContainerRegistryHostname: $INTERNAL_REGISTRY_URL
+    airGapContainerRegistryOrganization: 'eclipse'
+    nonProxyHosts: oauth-openshift.apps.$DOMAIN
+EOL
+
+# Start a golang workspace
```
Comment (tolusha) [inline .ci/oci-disconnected.sh:204]: Wrong comment location.

File: .ci/oci-disconnected.sh [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+initDefaults
+provisionOpenShiftOAuthUser
+
+# Deploy Eclipse Che
+chectl server:deploy --telemetry=off --k8spodwaittimeout=1800000 --che-operator-cr-patch-yaml=/tmp/che-cr-patch.yaml --che-operator-image=${INTERNAL_REGISTRY_URL}/eclipse/che-operator:nightly --platform=openshift --installer=operator
+
+provisionOAuth
+
+chectl auth:login -u admin -p admin
+chectl workspace:create --start --devfile="https://raw.githubusercontent.com/eclipse-che/che-devfile-registry/master/devfiles/go/devfile.yaml"
```
Comment (tolusha) [inline .ci/oci-disconnected.sh:214]: pls move devfile location to envar.

File: .ci/oci-disconnected.sh [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+# Copy Che Operator into private registry
+sudo skopeo copy --authfile=${REG_CREDS} --dest-tls-verify=false docker://quay.io/eclipse/che-operator:nightly docker://${INTERNAL_REGISTRY_URL}/eclipse/che-operator:nightly
+
+# OBTAIN PLUGINS
+IFS=$'\r\n' GLOBIGNORE='*' command eval 'PLUGINS_IMAGES=($(podman run --authfile=${XDG_RUNTIME_DIR}/containers/auth.json -it --rm \
+  --entrypoint cat ${INTERNAL_REGISTRY_URL}/eclipse/che-plugin-registry:nightly  /var/www/html/v3/external_images.txt))'
+for container in "${PLUGINS_IMAGES[@]}"
+do
+    :
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply .ci/oci-disconnected.sh:156]: @flacatus could you elaborate more on why do we need line with `:` here?

[the remark this answers is not in the document]
Comment (flacatus) [reply .ci/oci-disconnected.sh]: I remove it! Thanks for suggestion

File: .ci/oci-disconnected.sh (hunk 3 above)
Comment (flacatus) [reply .ci/oci-disconnected.sh]: No, my mistake. I removed it

File: .ci/oci-disconnected.sh (hunk 4 above)
Comment (flacatus) [reply .ci/oci-disconnected.sh:166]: Yes, I remove it

File: .ci/oci-disconnected.sh (hunk 6 above)
Comment (flacatus) [reply .ci/oci-disconnected.sh]: Done

**Review by mmorhun — APPROVED** [review]
LGTM

**Review by tolusha — APPROVED** [review]

**Pull request merged**